apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo/interceptor/OauthInterceptor.kt

```diff
@@ -0,0 +1,63 @@
+package com.apollographql.apollo.interceptor
+
+import com.apollographql.apollo.ApolloError
+import com.apollographql.apollo.ApolloException
+import com.apollographql.apollo.api.ApolloExperimental
+import com.apollographql.apollo.api.Response
+import com.apollographql.apollo.network.HttpExecutionContext
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.single
+
+@ApolloExperimental
+class OauthInterceptor(private val accessTokenProvider: AccessTokenProvider) : ApolloRequestInterceptor {
+  private fun <T> ApolloRequest<T>.withHeader(name: String, value: String): ApolloRequest<T> {
+    val httpRequestContext = executionContext[HttpExecutionContext.Request]
+        ?: HttpExecutionContext.Request(emptyMap())
+            .let {
+              it.copy(headers = it.headers + (name to value))
+            }
+
+    return ApolloRequest(
+        operation = operation,
+        scalarTypeAdapters = scalarTypeAdapters,
+        executionContext = executionContext + httpRequestContext
+    )
+  }
+
+  private suspend fun <T> proceedWithToken(request: ApolloRequest<T>, interceptorChain: ApolloInterceptorChain, token: String): Response<T> {
+    val newRequest = request.withHeader("Authorization", "Bearer $token")
+    return try {
+      interceptorChain.proceed(newRequest).single()
+    } catch (e: IllegalStateException) {
+      throw ApolloException(
+          message = "The downstream chain returned more than one response. Put OauthInterceptor just before the NetworkInterceptor",
+          error = ApolloError.Oauth
+      )
+    } catch (e: NoSuchElementException) {
+      throw ApolloException(
+          message = "The downstream chain did not return any response.",
+          error = ApolloError.Oauth
+      )
+    }
+  }
+
+  override fun <T> intercept(request: ApolloRequest<T>, interceptorChain: ApolloInterceptorChain): Flow<Response<T>> {
+    return flow {
+      val token = accessTokenProvider.currentAccessToken()
+
+      try {
+        val response = proceedWithToken(request, interceptorChain, token)
+        emit(response)
+      } catch (e: ApolloException) {
+        if (e.error == ApolloError.Network
+            && e.executionContext[HttpExecutionContext.Response]?.statusCode == 401) {
+          val newToken = accessTokenProvider.newAccessToken(token)
+          emit(proceedWithToken(request, interceptorChain, newToken))
+        } else {
+          throw e
+        }
+      }
```

Comment:
I was thinking of this:
```
val mutex = Mutex()
    return flow {
      mutex.withLock {
        val token = accessTokenProvider.currentToken()
        emit(token)
      }
    }.flatMapLatest { token ->
      proceedWithToken(request, interceptorChain, token)
    }.retry { error ->
      val shouldRenewToken = error is ApolloException && error.error is ApolloError.Http && error.error.statusCode == 401
      shouldRenewToken.also { renewToken ->
        if (renewToken) mutex.withLock {
          accessTokenProvider.renewToken()
        }
      }
    }
```

where `AccessTokenProvider`:
```
interface AccessTokenProvider {
  suspend fun currentToken(): String

  suspend fun renewToken()
}
```

What do you think?

Reply:
🤩 love it! This `retry` operator is really handy 👍 

About `renewToken`, I initially made it `suspend fun renewToken(previousToken: String)` to give a chance to the AccessTokenProvider to skip renewal if the token was just changed. 
It can be that multiple requests are blocked on the mutex and when the mutex is released, all of them will refresh the token, which is not really needed. So implementations could do something like this:

```
  override suspend fun renewToken(previousToken: String): String {
    if (previousToken != currentToken) {
      // Someone else renewed the token already
      return currentToken
    }
    
    // do the actual refresh and set currentToken
  }
```

Reply:
sure we can leave the signature as is:
```
override suspend fun renewToken(previousToken: String): String
```

Reply:
But then how do you make the previous token `flow` to the `retry` lambda ? That would mean forwarding that in a lot of different places no ?

Reply:
we can wrap it into own exception smth like:
```
return flow {
      val token = mutex.withLock { accessTokenProvider.currentToken() }
      emit(token)
    }.flatMapLatest { token ->
      proceedWithToken(request, interceptorChain, token).catch { error ->
                val oauthError = error is ApolloException && error.error is ApolloError.Http && error.error.statusCode == 401
           if (oauthError) {
              throw ApolloOuthException....
           }
       }
    }.retry { error ->
     ......
   }
```